src/inspector_socket.cc

```diff
@@ -0,0 +1,762 @@
+      payload_length64 <<= 8;
+      payload_length64 |= static_cast<unsigned char>(*p++);
+    }
+  }
+
+  size_t actual_masking_key_length = masked ? kMaskingKeyWidthInBytes : 0;
+  static const uint64_t max_payload_length = 0x7FFFFFFFFFFFFFFFull;
+  static size_t max_length = SIZE_MAX;
+  if (payload_length64 > max_payload_length ||
+      payload_length64 + actual_masking_key_length > max_length) {
+    // WebSocket frame length too large.
+    return FRAME_ERROR;
+  }
+  size_t payload_length = static_cast<size_t>(payload_length64);
+
+  size_t total_length = actual_masking_key_length + payload_length;
+  if (static_cast<size_t>(buffer_end - p) < total_length)
+    return FRAME_INCOMPLETE;
+
+  // Add 1 sizeof(char) for 0 terminator
+  *output = reinterpret_cast<char*>(malloc(payload_length + 1));
+  if (masked) {
+    const char* masking_key = p;
+    char* payload = const_cast<char*>(p + kMaskingKeyWidthInBytes);
+    for (size_t i = 0; i < payload_length; ++i)  // Unmask the payload.
+      (*output)[i] = payload[i] ^ masking_key[i % kMaskingKeyWidthInBytes];
+  } else {
+    strncpy(*output, p, payload_length);
+  }
+  (*output)[payload_length] = '\0';
+
+  size_t pos = p + actual_masking_key_length + payload_length - buffer_begin;
+  *bytes_consumed = pos;
+  return closed ? FRAME_CLOSE : FRAME_OK;
+}
+
+static void invoke_read_callback(inspector_socket_t* inspector,
+                                 int status, const uv_buf_t* buf) {
+  if (inspector->ws_state->read_cb) {
+    inspector->ws_state->read_cb(
+        reinterpret_cast<uv_stream_t*>(&inspector->client), status, buf);
+  }
+}
+
+static void shutdown_complete(inspector_socket_t* inspector) {
+  if (inspector->ws_state->close_cb) {
+    inspector->ws_state->close_cb(inspector, 0);
+  }
+  close_connection(inspector);
+}
+
+static void on_close_frame_written(uv_write_t* write, int status) {
+  inspector_socket_t* inspector =
+      reinterpret_cast<inspector_socket_t*>(write->handle->data);
+  write_request_cleanup(write, status);
+  inspector->ws_state->close_sent = true;
+  if (inspector->ws_state->received_close) {
+    shutdown_complete(inspector);
+  }
+}
+
+static void close_frame_received(inspector_socket_t* inspector) {
+  inspector->ws_state->received_close = true;
+  if (!inspector->ws_state->close_sent) {
+    invoke_read_callback(inspector, 0, 0);
+    write_to_client(inspector, CLOSE_FRAME, sizeof(CLOSE_FRAME),
+                    on_close_frame_written);
+  } else {
+    shutdown_complete(inspector);
+  }
+}
+
+static int parse_ws_frames(inspector_socket_t* inspector, size_t len) {
+  int bytes_consumed = 0;
+  char* output = nullptr;
+  bool compressed = false;
+
+  ws_decode_result r =  decode_frame_hybi17(inspector->buffer,
+                                            len, true /* client_frame */,
+                                            &bytes_consumed, &output,
+                                            &compressed);
```

Comment:
I suppose I have the same question here: why don't you write the result to a `std::vector<char>`?  That lets you neatly sidestep all the manual memory management and string juggling below.